curated_transformers/pipeline/distill.py

```diff
@@ -118,37 +127,49 @@ def distill(
             return losses
 
         student_hidden, student_backprop = self.model.begin_update(student_docs)
-
         # We have to pool the teacher output in the same way as the student.
         student_pooler = self.model.get_ref("tok2vec").get_ref("pooling")
-
         teacher_hidden = with_ragged_layers(student_pooler).predict(teacher_docs)
-
         layer_mapping = self._layer_mapping(teacher_hidden, student_hidden)
 
-        normalize = True
-        d_mse = []
-        for t_doc, s_doc in zip(teacher_hidden, student_hidden):
-            if normalize:
-                d_mse_doc = [
-                    (s_doc[student_layer] - t_doc[teacher_layer])
-                    / self.model.ops.xp.linalg.norm(t_doc[teacher_layer])
-                    for student_layer, teacher_layer in layer_mapping
-                ]
-            else:
-                d_mse_doc = [
-                    s_doc[student_layer] - t_doc[teacher_layer]
-                    for student_layer, teacher_layer in layer_mapping
-                ]
-
-            d_mse.append(d_mse_doc)
-
-        losses[self.name] += sum(
-            sum(float((d_layer**2).sum()) for d_layer in d) for d in d_mse
-        )
-
-        student_backprop(d_mse)
-        if sgd not in (None, False):
+        # Transpose the representations to the get the following shape: [layer, batch, seq, repr]
+        t_student_hidden = list(zip(*student_hidden))
+        t_teacher_hidden = list(zip(*teacher_hidden))
+
+        # Allocate loss accumulators on the GPU if possible.
+        cum_loss = self.model.ops.alloc1f(1)
+        layer_loss = self.model.ops.alloc1f(1)
+        t_mses = []
+        for student_layer_idx, teacher_layer_idx in layer_mapping:
+            student_batches = t_student_hidden[student_layer_idx]
+            teacher_batches = t_teacher_hidden[teacher_layer_idx]
```

Comment:
```suggestion
            student_batch = t_student_hidden[student_layer_idx]
            teacher_batch = t_teacher_hidden[teacher_layer_idx]
```

I think? Now that the layer is indexed, this is just a batch.

Reply:
Inlined the indexing op at the callsite instead.